internal/logging/journal.go

```diff
@@ -0,0 +1,108 @@
+package logging
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/coreos/go-systemd/v22/journal"
+	"go.etcd.io/etcd/client/pkg/v3/systemd"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"io"
+	"os"
+	"path/filepath"
+)
+
+// This file is written referring `logutil` in etcd pkg
+// Reference: https://github.com/etcd-io/etcd/blob/946a5a6f25c3b6b89408ab447852731bde6e6289/client/pkg/logutil/zap_journal.go
+type journalWriter struct {
+	io.Writer
+}
+
+// newJournalWriter wraps "io.Writer" to redirect log output
+// to the local systemd journal.
+func newJournalWriter(wr io.Writer) (io.Writer, error) {
+	return &journalWriter{Writer: wr}, systemd.DialJournal()
+}
+
+// stringToZapLevel maps level in string to zapcore.Level
+var stringToZapLevel = func() map[string]zapcore.Level {
+	logLvl := make(map[string]zapcore.Level)
+
+	for level, Value := range map[string]zapcore.Level{
+		"debug":  zap.DebugLevel,
+		"info":   zap.InfoLevel,
+		"warn":   zap.WarnLevel,
+		"error":  zap.ErrorLevel,
+		"fatal":  zap.FatalLevel,
+		"panic":  zap.PanicLevel,
+		"dpanic": zap.DPanicLevel,
+	} {
+		logLvl[level] = Value
+	}
+	return logLvl
+}()
+
+// logLine reads the logger fields which are used to construct and send log messages to journal
+type logLine struct {
+	Level      string `json:"level"`
+	LoggerName string `json:"logger"`
+	Caller     string `json:"caller"`
+	Message    string `json:"msg"`
+	Stack      string `json:"stacktrace"`
+}
+
+func (w *journalWriter) Write(p []byte) (int, error) {
+	//var line map[string]string
+	line := &logLine{}
+	//m := json.NewDecoder(bytes.NewReader(p)).Buffered()
+	if err := json.NewDecoder(bytes.NewReader(p)).Decode(line); err != nil {
+		panic(err)
+	}
+
+	var pri journal.Priority
+	lvl := stringToZapLevel[line.Level]
+	switch lvl {
+	case zapcore.DebugLevel:
+		pri = journal.PriDebug
+	case zapcore.InfoLevel:
+		pri = journal.PriInfo
+
+	case zapcore.WarnLevel:
+		pri = journal.PriWarning
+	case zapcore.ErrorLevel:
+		pri = journal.PriErr
+
+	case zapcore.DPanicLevel:
+		pri = journal.PriCrit
+	case zapcore.PanicLevel:
+		pri = journal.PriCrit
+	case zapcore.FatalLevel:
+		pri = journal.PriCrit
+
+	default:
+		panic(fmt.Errorf("unknown log level: %q", line.Level))
+	}
+
+	consoleEncEntry := zapcore.Entry{
+		Level:      stringToZapLevel[line.Level],
+		LoggerName: line.LoggerName,
+		Message:    line.Message,
+		Stack:      line.Stack,
+	}
+
+	message, err := zapcore.NewConsoleEncoder(defaultConfig.EncoderConfig).EncodeEntry(consoleEncEntry, nil)
```

Comment:
Make no sense to create an encoder for every message. Pass the encoder config to `NewJournalWriter` and create an encoder from that.